File: siaqodb/Documents/Document.cs
using Sqo.Exceptions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

namespace Sqo.Documents
{
    [System.Reflection.Obfuscation(Exclude = true)]
    public class Document
    {
        public Document()
        {
        }
        internal Document(string key, byte[] content)
        {
            this.Key = key;
            this.content = content;
        }
       
        internal bool ShouldSerializeIsDirty()
        {
            return false;
        }
        private string key;

        public string Key
        {
            get
            {
                return this.key;
            }
            set
            {
                key = value;
            }
        }
        private byte[] content;

        public byte[] Content
        {
            get { return content; }
            set { content = value; }
        }
        public string Version { get; set; }
        public void SetContent<T>(T objValue)
        {
            this.SetContent((object)objValue);
        }
        public void SetContent(object objValue)
        {

            if (objValue == null)
                throw new ArgumentNullException("objValue");

            if (this.Key == null && SiaqodbConfigurator.KeyConventions.ContainsKey(objValue.GetType()))
            {
                this.Key = SiaqodbConfigurator.KeyConventions[objValue.GetType()](objValue);
            }
            if (this.Version == null && SiaqodbConfigurator.VersionSetConventions.ContainsKey(objValue.GetType()))
            {
                this.Version = SiaqodbConfigurator.VersionSetConventions[objValue.GetType()](objValue);
            }
            byte[] serializedObj = SiaqodbConfigurator.DocumentSerializer.Serialize(objValue);
            this.Content = serializedObj;
            
        }
        public object GetContent( Type type)
        {
            if (this.Content == null || this.Content.Length == 0)
                return null;
        
            object obj = SiaqodbConfigurator.DocumentSerializer.Deserialize(type, this.Content);
            if (SiaqodbConfigurator.VersionGetConventions.ContainsKey(type))
            {
                SiaqodbConfigurator.VersionGetConventions[type](obj, this.Version);
            }
            return obj;
        }
        public T GetContent<T>()
        {
            return (T)((object)this.GetContent(typeof(T)));
        }




        public void SetTag<T>(string tagName, T value)
        {
            this.SetTag(tagName, (object)value);

        }
        public void SetTag(string tagName, object value)
        {
            tagName = tagName.ToLower();
            Type type = value.GetType();
            if (!ValidTagName(tagName))
            {
                throw new SiaqodbException("Tag name:" + tagName + " is not valid.");
            }
            if (Tags == null)
                Tags = new Dictionary<string, object>();
            if (type == typeof(int) || type == typeof(long))
            {
                Tags[tagName] = Convert.ToInt64(value);
            }
            else if (type == typeof(double) || type == typeof(float))
            {
                Tags[tagName] = Convert.ToDouble(value);
            }
            else if (type == typeof(DateTime) || type == typeof(string) || type == typeof(bool))
            {

                Tags[tagName] = value;
            }

            else
            {
                throw new SiaqodbException("Tag type:" + type.ToString() + " not supported.");
            }
        }

        Dictionary<string, object> tags;
        public Dictionary<string, object> Tags
        {
            get
            {
                return tags;
            }
            set
            {
                tags = value;

            }
        }

        internal bool ValidTagName(string tagName)
        {

            if (!Regex.IsMatch(tagName, "^[a-zA-Z][0-9a-zA-Z_-]*$"))
            {
                return false;
            }
            return true;
        }

        public T GetTag<T>(string tagName)
        {
            if (Tags != null)
            {
                tagName = tagName.ToLower();
                if (Tags.ContainsKey(tagName))
                {
                    if (Tags[tagName].GetType() != typeof(T))
                    {
                        return (T)Convert.ChangeType(Tags[tagName], typeof(T));
                    }
                    else
                    {
                        return (T)Tags[tagName];
                    }
                }
            }
            return default(T);
        }


    }
}

File: Android/sqo/src/main/java/dotissi/sqo/Document.java
package dotissi.sqo;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.HashMap;
import java.util.Objects;


public class Document {
    public String key;
    public byte[] content;
    public String version;
    public HashMap<String, Object> tags;
    public HashMap<String, Object> getTags() {
        return tags;
    }
    public void setTags(HashMap<String, Object> tags) {
        this.tags = tags;
    }
    public String getKey() {
        return key;
    }
    public void setKey(String key) {
        this.key = key;
    }
    public String getVersion() {
        return version;
    }
    public void setVersion(String version) {
        this.version = version;
    }
    public byte[] getRawContent() {
        return content;
    }
    public void setRawContent(byte[] content) {
        this.content = content;
    }
    public Object getContent(Class<? extends Object> type) {
        if (this.content == null || this.content.length == 0)
            return null;
        Object obj = SiaqodbConfigurator.getSerializer().deserialize(type, this.content);
        //TODO
        /*IVersionConvention versionConvention=CryptonorConfigurator.getVersionConvention();
        if(versionConvention!=null)
        {
            versionConvention.setVersion(this.getVersion(),obj);
        }*/
        return obj;
    }
    public void setContent(Object content) {
          if (content == null)
            throw new IllegalArgumentException("content");

       //TODO key conventions
        /*IKeyConvention keyConvention=CryptonorConfigurator.getKeyConvention();
        if(this.key==null && keyConvention!=null)
        {
            this.key=keyConvention.getKey(objValue);
        }
        IVersionConvention versionConvention=CryptonorConfigurator.getVersionConvention();
        if(this.version==null && versionConvention!=null)
        {
            this.version= versionConvention.getVersion(objValue);
        }*/
        byte[] serializedObj = SiaqodbConfigurator.getSerializer().serialize(content);
        this.content = serializedObj;

    }
    public void setTag(String tagName, Object value) throws SiaqodbException {
            tagName = tagName.toLowerCase();
        Class<? extends Object> type = value.getClass();
        if (tags == null)
            tags = new HashMap<String, Object>();
        if (type == int.class || type == long.class || type == Long.class || type == Integer.class)
        {
            tags.put(tagName,Long.parseLong(value.toString()));
        }
        else if ( type == float.class|| type == Float.class || type == double.class ||type == Double.class)
        {
            tags.put(tagName,Double.parseDouble(value.toString()));
        }

        else if ( type == String.class
                || type == boolean.class|| type == Boolean.class || type==Date.class)
        {
            tags.put(tagName,value);
        }

        else
        {
            throw new SiaqodbException("Tag type:" + type.toString() + " not supported.");
        }
    }
    public Object getTag(String tagName,Class<? extends Object> type) throws ParseException
    {
        if (tags != null)
        {
            tagName = tagName.toLowerCase();
            if (tags.containsKey(tagName))
            {
                if (tags.get(tagName).getClass() != type)
                {
                    if(type == Date.class){
                        SimpleDateFormat myDateFormat= new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss");

                        return  myDateFormat.parse(tags.get(tagName).toString());
                    }
                    else if(type == int.class || type == Integer.class ){
                        return type.cast(Integer.parseInt( tags.get(tagName).toString()));
                    }
                    else if(type == float.class || type == Float.class ){
                        return type.cast(Float.parseFloat( tags.get(tagName).toString()));
                    }
                    return type.cast(tags.get(tagName));
                }
                else
                {
                    return tags.get(tagName);
                }
            }
        }
        return null;
    }
}
